main/src/library/DecodingReader.flix

```diff
@@ -0,0 +1,123 @@
+
+    pub def read(buffer: Array[Char, rb], reader: DecodingReader[t, rr]): Result[IoError, Int32] \ rb + rr + Readable.Aef[t] + IO = DecodingReader.read(buffer, reader)
+
+}
+
+mod DecodingReader {
+    import java.nio.charset.CharsetDecoder
+    import java.nio.charset.CodingErrorAction
+    import java.nio.ByteBuffer
+    import java.nio.CharBuffer
+
+    ///
+    /// Wraps `byteReader` in a `DecodingReader`.
+    ///
+    pub def wrap(rc: Region[rr], charset: CharacterSet, byteReader: t): DecodingReader[t, rr] \ rr + IO =
+        let CharacterSet.CharacterSet(cs) = charset;
+        new DecodingReader @ rc {
+            rc = rc,
+            decoder = cs.newDecoder().onMalformedInput(CodingErrorAction.REPLACE),
+            byteReader = byteReader,
+            leftovers = Array.empty(rc, 0)
+        }
+
+    ///
+    /// Reads bytes from the underlying byteReader and decodes the result into `dest`.
+    ///
+    /// Returns `Ok(k)` to signify that `k` chars were successfully read and written to `dest`.
+    ///
+    /// Guarantees that `0 <= k <= length(dest)`.
+    ///
+    /// Returns `Err(e)` if some underlying error occurs.
+    ///
+    pub def read(dest: Array[Char, rb], reader: DecodingReader[t, rr]): Result[IoError, Int32] \ IO + rr + rb + Readable.Aef[t] with Readable[t] where Readable.Elm[t] ~ Int8 =
+        forM(
+            (src, endOfInput) <- {
+                // At most 4 bytes per char, guarantees we can fill dest if there are enough bytes to do so
+                let totalBytesNeeded = 4 * Array.length(dest);
+                readBytes(totalBytesNeeded, reader)
+            };
+            charsRead <- {
+                let charsRead = decode(src, dest, endOfInput, reader);
+
+                // charsRead == 0 would imply end of input to the caller.
+                // block again until end of input or we have at least one char to return
+                if (not endOfInput and charsRead == 0) read(dest, reader) else Ok(charsRead)
+            }
+        ) yield charsRead
+
+    ///
+    /// Returns `Ok(bytes, endOfInput)` where `bytes` are the next bytes to be decoded.
+    ///
+    /// `bytes` contains `leftovers` from the last read prepended to new bytes read from the `byteReader` (if necessary).
+    ///
+    /// `length(bytes) == numBytes` if `endOfInput` is false. `length(bytes) <= numBytes` otherwise.
+    ///
+    def readBytes(numBytes: Int32, reader: DecodingReader[t, r]): Result[IoError, (Array[Int8, r], Bool)] \ r + Readable.Aef[t] with Readable[t] where Readable.Elm[t] ~ Int8 =
+        let bytesToRead = numBytes - Array.length(reader->leftovers);
+        if (bytesToRead > 0) {
+            let src = Array.empty(reader->rc, bytesToRead);
+            forM(
+                bytesRead <- Readable.read(src, reader->byteReader)
+            ) yield {
+                let newBytes = Array.take(reader->rc, bytesRead, src);
+                let bytes = Array.append(reader->rc, reader->leftovers, newBytes);
```

Comment:
I wonder if there's something to be done about reusing existing arrays instead of these linear-tie operations.

But I don't think we should worry about that now.